### PR TITLE
feat: Adding typed sign support for NFT permit

### LINKED ISSUE #27396: Support ERC-721 Permit Signatures
### What is this about?

Add support for ERC-721 Permits [(EIP-4494)](https://eips.ethereum.org/EIPS/eip-4494)

Specs:
- todo (Title, description, row specifications)
- Simulation to be addressed: https://github.com/MetaMask/metamask-extension/issues/27394

### Related Issues

Related: https://github.com/MetaMask/MetaMask-planning/issues/2936 (NFT Approve w/ Simulation)

### Repro

```
   await window.ethereum.request({ "method": "eth_signTypedData_v4", "params": [
  "0x9B285340bB67eb56fd093a612784c0B5FD5304c4", 
 {
            domain: {
              name: 'Uniswap V3 Positions NFT-V1',
              version: '1',
              chainId: 1,
              verifyingContract: '0xC36442b4a4522E871399CD717aBDD847Ab11FE88',
            },
            types: {
              Permit: [
                { name: 'spender', type: 'address' },
                { name: 'tokenId', type: 'uint256' },
                { name: 'nonce', type: 'uint256' },
                { name: 'deadline', type: 'uint256' },
              ],
            },
            primaryType: 'Permit',
            message: {
              spender: '0x00000000Ede6d8D217c60f93191C060747324bca',
              tokenId: '3606393',
              nonce: '0',
              deadline: '1734995006',
            },
          }], });
```

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [ ] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by jpuri**
## **Description**

Adding support for NFT permit signature request.

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/27396

## **Manual testing steps**

1. Submit an NFT permit request
2. Check the confirmation page that appears

## **Screenshots/Recordings**
Screenshot 2024-10-11 at 7 43 59 PM

## **Pre-merge author checklist**

- [X] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import { parseTypedDataMessage } from '../../../../shared/modules/transaction.utils';
+import { TokenStandard } from '../../../../shared/constants/transaction';
+import { MESSAGE_TYPE } from '../../../../shared/constants/app';
+import { TypedSignSignaturePrimaryTypes } from '../constants';
+
+export const useTypedSignSignatureInfo = (
+  confirmation: SignatureRequestType,
+) => {
+  if (!confirmation) {
+    return {};
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts:18]: Minor, would it be more explicit to return `undefined` here so we can insist primary type and standard are always returned?

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  if (
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts:20]: Minor, could this be combined with the above condition so we have a single guard with all applicable conditions?

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.test.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { renderHook } from '@testing-library/react-hooks';
+
+import { TokenStandard } from '../../../../shared/constants/transaction';
+import { permitNFTSignatureMsg } from '../../../../test/data/confirmations/typed_sign';
+import { TypedSignSignaturePrimaryTypes } from '../constants';
+import { useTypedSignSignatureInfo } from './useTypedSignSignatureInfo';
+
+describe('useTypedSignSignatureInfo', () => {
+  describe('isNftTransfer', () => {
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/hooks/useTypedSignSignatureInfo.test.js:9]: Was this nested describe accidental?

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.test.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    it('should return false if transaction is not NFT transfer', () => {
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/hooks/useTypedSignSignatureInfo.test.js:10]: Minor, is it worth unit tests for the early return conditions?

**Review by matthewwalsh0 — DISMISSED** [review]

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.test.js (hunk 4 above)
Comment (jpuri) [reply ui/pages/confirmations/hooks/useTypedSignSignatureInfo.test.js]: yep, I added one test for that

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.test.js (hunk 3 above)
Comment (jpuri) [reply ui/pages/confirmations/hooks/useTypedSignSignatureInfo.test.js]: yep I removed it, thanks

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts (hunk 2 above)
Comment (jpuri) [reply ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts:17]: yep done

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts (hunk 1 above)
Comment (jpuri) [reply ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts]: I do not get this point exactly, I return empty object here for ease of destructuring the result.

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  }, [confirmation]);
+
+  // here we are using presence of tokenId in typed message data to know if its NFT permit
+  // we can get contract details for verifyingContract but that is async process taking longer
+  // and result in confirmation page content loading late
+  const tokenStandard = useMemo(() => {
+    if (primaryType !== TypedSignSignaturePrimaryTypes.PERMIT) {
+      return undefined;
+    }
+    console.log(confirmation, confirmation?.msgParams?.data);
```
Comment (matthewwalsh0) [inline ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts:40]: Accidental?

**Review by matthewwalsh0 — DISMISSED** [review]

File: ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts (hunk 5 above)
Comment (jpuri) [reply ui/pages/confirmations/hooks/useTypedSignSignatureInfo.ts]: Removed

**Comment by pedronfigueiredo** [comment]
Would it make sense to add an e2e test?

**Comment by jpuri** [comment]
@pedronfigueiredo : yep that is good idea, let me do that.

**Review by matthewwalsh0 — APPROVED** [review]

**Review by pedronfigueiredo — APPROVED** [review]

**Pull request merged**